docs/cryptodoc/src/05_06_hss_lms.rst

```diff
@@ -0,0 +1,422 @@
+are given. The classes ``HSS_LMS_PublicKeyInternal`` and
+``HSS_LMS_PrivateKeyInternal`` realize the public and secret key, respectively.
+
+Botan's HSS/LMS implementation derives LMS seeds and identifiers
+by the same method Cisco's reference implementation applies. This approach
+is called ``SECRET_METHOD 2`` in the Cisco implementation's configuration.
+``SEED`` and ``I`` of child LMS trees are derived from the values of their
+parents and their position in the hypertree. This operation is similar to the
+pseudorandom key generation method of [RFC8554]_ Appendix A.
+The derivation functions are the following:
+
+.. math::
+   \mathtt{SEED_{child}}\ &\mathtt{= Hash(I_{parent}\ ||\ u32str(q_{parent})\
+   ||\ u16str(0xfffe)\ ||\ u8str(0xff)\ ||\ SEED_{parent})}
+
+   \mathtt{I_{child}}\    &\mathtt{= Hash(I_{parent}\ ||\ u32str(q_{parent})\
+   ||\ u16str(0xffff)\ ||\ u8str(0xff)\ ||\ SEED_{parent})}
+
+:math:`\mathtt{I_{parent}}` and :math:`\mathtt{SEED_{parent}}` are the
+identifier and seed of the parent LMS tree, while :math:`\mathtt{I_{child}}`
+and :math:`\mathtt{SEED_{child}}` are derived for the new child.
+:math:`\mathtt{q_{parent}}` is the parent's LMS leaf index used to sign the
+child LMS tree's public key. Note that since the third entry separates their
+domain, the hash inputs will never collide with
+the ones of the LM-OTS computations of ``x[i]`` and ``C``.
+
+The method ``HSS_LMS_Signature_Operation::sign`` is used for signature creation,
+implementing Algorithm 8 of [RFC8554]_.
+``HSS_LMS_Verification_Operation::is_valid_signature`` provides signature
+verification, as in Section 6.3. of [RFC8554]_.
+
+.. _pubkey/hss_lms/key_gen:
+
+Key Generation
+--------------
+
+HSS/LMS key generation follows Section 6.1. of [RFC8554]_ and is implemented
+within the ``HSS_LMS_PrivateKeyInternal`` constructor and
+``HSS_LMS_PublicKeyInternal::create``. It works as follows:
+
+.. admonition:: HSS/LMS Key Generation
+
+   **Input:**
+
+   -  ``rng``: random number generator
+   -  ``L``: The number of levels in the HSS/LMS hypertree
+   -  ``lms-params[0], ..., lms-params[L-1]`` : LMS parameter sets at all
+      ``L`` levels
+   -  ``lm-ots-params[0], ..., lm-ots-params[L-1]``: LM-OTS parameter sets at all
+      ``L`` levels
+
+   **Output:**
+
+   -  ``SK``, ``PK``: secret and public key
+
+   **Steps:**
+
+   1. Generate new values ``SEED`` and ``I`` using ``rng``.
+   2. ``idx = 0`` sets the initial signature index.
+   3. | ``SK = {L, idx, lms-params[0], lm-ots-params[0], ..., lms-params[L-1],``
+      |       ``lm-ots-params[L-1], SEED, I}``.
+   4. Construct the root LMS secret key ``lms-sk[0]`` with parameters
+      ``lms-params[0]`` and ``lm-ots-params[0]`` containing ``SEED`` and ``I``
+      (see :ref:`LMS <pubkey/hss_lms/lms>`).
+   5. Construct the root LMS public key ``lms-pk[0]`` from ``lms-sk[0]``
+      (see :ref:`LMS <pubkey/hss_lms/lms>`).
+   6. ``PK = {L, lms-pk[0]}``.
+
+   **Notes:**
```

Comment:
Should the notes also reflect that the given parameters for lm-ots and lms for the different levels are checked concerning the restrictions imposed? i.e. matching hash function etc.